### PR TITLE
Bumping hlsjs version

### PR DESCRIPTION
**Pull request by ruslandinov**
PT story https://www.pivotaltracker.com/story/show/135118117

Made the wrapper compatible with hls.js v0.6.12. So, now it supports hls.js versions from v0.5.46 to v0.6.12.
Breaking changes were introduced starting hls.js v0.6.2, so Idifferent fragment loader class was introduced in wrapper to handle that.

**Bundled hls.js version wasn't changed, it's still v0.5.46**

### CODE REVIEW

File: lib/integration/p2p-loader-generator.js [hunk 1]
```diff
@@ -1,213 +1,21 @@
...
+    let hlsjsVersion = hlsjsWrapper.hls.constructor.version;
+    if (hlsjsVersion === undefined) {
+        return p2pLoaderGenerator_v0_5_46(hlsjsWrapper);
+    } else {
+        let versionParts = hlsjsVersion.split('.');
+        let minor = versionParts[1];
+        let patch = versionParts[2];
+        if (Number(minor) >= 6 && Number(patch) >= 2) {
+            return p2pLoaderGenerator_v0_6_2(hlsjsWrapper);
+        } else {
```
Comment (jlouazel) [inline lib/integration/p2p-loader-generator.js]: Useless else, you can put the statement `p2pLoaderGenerator_v0_5_46` directly below the if.

File: package.json [hunk 2]
```diff
@@ -59,6 +59,7 @@
     "karma": "^0.13.22",
     "karma-chrome-launcher": "^1.0.1",
```
Comment (jlouazel) [inline package.json]: Maybe it's time to update karma/mocha deps? FYI here is the current status:

Package |  Current | Latest
--- | --- | ---
karma | 0.13.22 | 1.3.0
karma-chrome-launcher | 1.0.1 | 2.0.0
mocha | 2.5.3 | 3.2.0                               

Time to update? 🙃

Comment (jlouazel) [inline package.json]: Also, I don't know why we use `should` instead of `chai`. Like my comment above, this is not the point here but we should think about for future PRs.

**Comment by jlouazel** [comment]
I don't see tests about returning a loader depending on what you get from the parsing. Talking about that, I'm assuming you will fallback to the 0.5.46 loader in case the parsing gone wrong. Am I right?

**Comment by jlouazel** [comment]
Otherwise, all of this seems good to me 👌

File: lib/integration/p2p-loader-generator-v0-6-2.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                throw new Error('P2P loader was not reset correctly, internal state indicates unfinalized request');
+            }
+
+            let url = this.context.url;
+            let xhrSetup = this.xhrSetup;
+            let {headers, withCredentials} = extractInfoFromXhrSetup(xhrSetup, url);
+
+            if (this.byteRange) {
+                headers.Range = `bytes=${this.context.frag.byteRangeStartOffset}-${this.context.frag.byteRangeEndOffset - 1}`;
+            }
```
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Duplicates with https://github.com/streamroot/hlsjs-p2p-wrapper/pull/96/files#diff-a3cf0874d55f2155fdfbb36f0856f554R61
It seems it was already the case in the v0.5.46. It would be good to only compute this range in one place

File: lib/integration/p2p-loader-generator-v0-6-2.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                this.stats.retry++;
+                this.reset(false);
+            } else {
+                console.error(`${status} while loading ${this.context.url}` );
+                let xhrEventShim = {
+                    target: {
+                        status,
+                    },
+                };
+                this.onError(xhrEventShim);
```
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Be careful, the `onError` signature has changed as well !!
https://github.com/dailymotion/hls.js/blob/master/src/utils/xhr-loader.js#L113

**Comment by ruslandinov** [comment]
@jlouazel regarding this https://github.com/streamroot/hlsjs-p2p-wrapper/pull/96#issuecomment-265117379 comment: version getter was introduced in hls.js starting from v0.6.1, so yes, if it's undef, I fallback to old loader

File: lib/integration/p2p-loader-generator-v0-6-2.js [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            this.peerAgentLoader = null;
+        }
+
+        load(context, config, callbacks) {
+            if (!callbacks.onProgress) {
+                throw new Error('P2P loader expects progress-callback to be passed for ABR stats (use only as `fLoader` in config)');
+            }
+
+            if (!context.frag) {
+                throw new Error('P2P loader can only be used for media fragments (use only as `fLoader` in config)');
```
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Good catch, thank you. 
I'm not sure that we need this check at all with current implementation. We don't expose our custom loader creation method outside wrapper, so I'd rather wipe this check out. Otherwise, I can check `if (context.hasOwnProperty('progressData'))`, which is defined in https://github.com/dailymotion/hls.js/blob/master/src/loader/fragment-loader.js#L43. But I think that this check is unnecessary complexity.

Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: What do you think @AxelDelmas ?

Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Nevermind the question above, looks like this check is more than enough https://github.com/streamroot/hlsjs-p2p-wrapper/pull/96/files/41e2b84eebeec0c9f55cf2422cf9f8414b5c2b99#diff-a3cf0874d55f2155fdfbb36f0856f554R47 to enusre, that loader is used as `fLoader`, because `onProgress` callback is defined in fragment loader only.

File: lib/integration/p2p-loader-generator-v0-6-2.js [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+            }
+
+            if (!isNaN(context.byteRangeStartOffset) && !isNaN(context.byteRangeEndOffset)) {
+                this.byteRange = context.byteRangeStartOffset + '-' + context.byteRangeEndOffset;
+            }
+
+            this.context = context;
+            // this.frag = context.frag;
+            // this.url = context.url;
+            // this.responseType = context.responseType;
```
Comment (jlouazel) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: 👍

**Comment by jlouazel** [comment]
@ruslandinov Got it

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 5 above)
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Yep should be enough. True that if we don't expose our custom loader outside the wrapper, we have actually no reason to make those checks.

File: package.json (hunk 2 above)
Comment (AxelDelmas) [inline package.json]: Let's do this in a separate PR

File: lib/integration/p2p-loader-generator.js (hunk 1 above)
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator.js]: D'accord, fixed.

File: package.json (hunk 2 above)
Comment (ruslandinov) [inline package.json]: I've updated the deps.

File: lib/integration/p2p-loader-generator-v0-6-2.js [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            }
+
+            if (!hlsjsWrapper.peerAgentModule) {
+                // Should not happen. Means we loaded a frag before the manifest, of there's a problem in the dispose sequence.
+                throw new Error('Peer agent is not existing yet');
+            }
+
+            if (!isNaN(context.byteRangeStartOffset) && !isNaN(context.byteRangeEndOffset)) {
+                this.byteRange = context.byteRangeStartOffset + '-' + context.byteRangeEndOffset;
+            }
```
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: My bad, fixed.

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 6 above)
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: My bad, fixed.

File: lib/integration/p2p-loader-generator-v0-6-2.js [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+            this.context = context;
+            // this.frag = context.frag;
+            // this.url = context.url;
+            // this.responseType = context.responseType;
+            this.onSuccess = callbacks.onSuccess;
+            this.onProgress = callbacks.onProgress;
+            this.onTimeout = callbacks.onTimeout;
+            this.onError = callbacks.onError;
```
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Done.

File: lib/integration/p2p-loader-generator-v0-6-2.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            this.stats = { trequest: performance.now(), retry: 0 };
+            this.timeout = config.timeout;
+            this.maxRetry = config.maxRetry;
+            this.retryDelay = config.retryDelay;
```
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Done, except that I kept `this.retryDelay` -- it's value is modifed due to exponential backoff, wanted to be sure we don't change hls.js config value.

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 4 above)
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Fixed.

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 3 above)
Comment (ruslandinov) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: Simplified it both in v0.5.46 & v0.6.2 loaders.

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 5 above)
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: :+1:

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 7 above)
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: :+1:

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 6 above)
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: :+1:

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 9 above)
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: :+1:

File: lib/integration/p2p-loader-generator-v0-6-2.js (hunk 3 above)
Comment (AxelDelmas) [inline lib/integration/p2p-loader-generator-v0-6-2.js]: :+1:
